File: backend/routes/pre-leads.js
const express = require('express');
const router = express.Router();
const { getDb } = require('../config/db');
const { protect, authorize } = require('../middleware/auth');
const { ObjectId } = require('mongodb');
const multer = require('multer');

const imgUpload = multer({
  storage: multer.memoryStorage(),
  limits: { fileSize: 10 * 1024 * 1024 },
  fileFilter: (req, file, cb) => {
    const ok = file.mimetype.startsWith('image/');
    cb(ok ? null : new Error('Solo se permiten imágenes'), ok);
  }
});

// Roles que pueden gestionar leads en "Nuevo Lead"
const isProcesamiento = (role) => {
  const r = String(role || '').toLowerCase().trim();
  return ['admin','administrador','administrator'].includes(r) || r.startsWith('procesamiento');
};

// Genera un ID de 5 dígitos único
async function generarLeadId(db) {
  const max = 99999;
  const min = 10000;
  let id;
  let intentos = 0;
  do {
    id = String(Math.floor(Math.random() * (max - min + 1)) + min);
    const existe = await db.collection('pre_leads').findOne({ leadId: id });
    if (!existe) break;
    intentos++;
  } while (intentos < 20);
  return id;
}

// POST /api/pre-leads — crear nuevo lead (cualquier usuario autenticado)
router.post('/', protect, async (req, res) => {
  try {
    const db = getDb();
    if (!db) return res.status(500).json({ success: false, message: 'DB no disponible' });

    const {
      nombre, correo, phone1, phone2, direccion,
      fechaNacimiento, servicio, mercado, nota,
      agenteUsername, agenteName
    } = req.body;

    if (!nombre || !phone1 || !direccion || !servicio || !fechaNacimiento || !correo) {
      return res.status(400).json({ success: false, message: 'Faltan campos obligatorios' });
    }

    const leadId = await generarLeadId(db);
    const now = new Date();

    const doc = {
      leadId,
      nombre:          String(nombre).trim(),
      correo:          String(correo).trim(),
      phone1:          String(phone1).trim(),
      phone2:          String(phone2 || '').trim(),
      direccion:       String(direccion).trim(),
      fechaNacimiento: String(fechaNacimiento).trim(),
      servicio:        String(servicio).trim(),
      mercado:         String(mercado || '').trim(),
      nota:            String(nota || '').trim(),
      agenteUsername:  req.user.username || String(agenteUsername || '').trim(),
      agenteName:      req.user.name || String(agenteName || '').trim(),
      status:          null,
      notaProcesamiento: '',
      fechaVenta:      '',
      fechaInstalacion: '',
      resolucion:      null,
      creadoEn:        now,
      actualizadoEn:   now,
    };

    const result = await db.collection('pre_leads').insertOne(doc);
    doc._id = result.insertedId;

    res.status(201).json({ success: true, lead: doc });
  } catch (err) {
    console.error('[PRE-LEADS] POST error:', err);
    res.status(500).json({ success: false, message: 'Error interno del servidor' });
  }
});

// GET /api/pre-leads — todos los leads (solo Admin/Procesamiento, filtrado por mercado)
router.get('/', protect, async (req, res) => {
  try {
    if (!isProcesamiento(req.user.role)) {
      return res.status(403).json({ success: false, message: 'Acceso denegado' });
    }

    const db = getDb();
    if (!db) return res.status(500).json({ success: false, message: 'DB no disponible' });

    const roleNorm = String(req.user.role || '').toLowerCase().trim();
    const isAdmin  = ['admin', 'administrador', 'administrator'].includes(roleNorm);

    // Filtro de mercado según el tipo de procesamiento
    let query = {};
    if (!isAdmin) {
      if (roleNorm.includes('icon')) {
        query.mercado = { $regex: /^icon$/i };
      } else if (roleNorm.includes('bamo')) {
        query.mercado = { $regex: /^bamo$/i };
      }
      // 'procesamiento' genérico ve todos
    }

    const leads = await db.collection('pre_leads')
      .find(query)
      .sort({ creadoEn: -1 })
      .toArray();

    res.json({ success: true, leads: leads.map(l => ({ ...l, _id: String(l._id) })) });
  } catch (err) {
    console.error('[PRE-LEADS] GET error:', err);
    res.status(500).json({ success: false, message: 'Error interno del servidor' });
  }
});

// GET /api/pre-leads/mis-leads — leads propios del agente
router.get('/mis-leads', protect, async (req, res) => {
  try {
    const db = getDb();
    if (!db) return res.status(500).json({ success: false, message: 'DB no disponible' });

    const leads = await db.collection('pre_leads')
      .find({ agenteUsername: req.user.username })
      .sort({ creadoEn: -1 })
      .toArray();

    res.json({ success: true, leads: leads.map(l => ({ ...l, _id: String(l._id) })) });
  } catch (err) {
    console.error('[PRE-LEADS] GET mis-leads error:', err);
    res.status(500).json({ success: false, message: 'Error interno del servidor' });
  }
});

// PUT /api/pre-leads/:id — actualizar campos del lead (Admin/Procesamiento)
router.put('/:id', protect, async (req, res) => {
  try {
    if (!isProcesamiento(req.user.role)) {
      return res.status(403).json({ success: false, message: 'Acceso denegado' });
    }

    const db = getDb();
    if (!db) return res.status(500).json({ success: false, message: 'DB no disponible' });

    let oid;
    try { oid = new ObjectId(req.params.id); } catch {
      return res.status(400).json({ success: false, message: 'ID inválido' });
    }

    const allowed = [
      'status', 'notaProcesamiento', 'fechaVenta', 'fechaInstalacion',
      'nombre', 'correo', 'phone1', 'phone2', 'direccion',
      'fechaNacimiento', 'servicio', 'mercado', 'nota'
    ];
    const update = {};
    allowed.forEach(k => {
      if (req.body[k] !== undefined) update[k] = String(req.body[k]).trim();
    });

    if (Object.keys(update).length === 0) {
      return res.status(400).json({ success: false, message: 'Sin campos para actualizar' });
    }

    update.actualizadoEn = new Date();

    const result = await db.collection('pre_leads').updateOne(
      { _id: oid },
      { $set: update }
    );

    if (result.matchedCount === 0) {
      return res.status(404).json({ success: false, message: 'Lead no encontrado' });
    }

    res.json({ success: true });
  } catch (err) {
    console.error('[PRE-LEADS] PUT error:', err);
    res.status(500).json({ success: false, message: 'Error interno del servidor' });
  }
});

// PUT /api/pre-leads/:id/resolver — marcar como completada o pendiente
router.put('/:id/resolver', protect, async (req, res) => {
  try {
    if (!isProcesamiento(req.user.role)) {
      return res.status(403).json({ success: false, message: 'Acceso denegado' });
    }

    const db = getDb();
    if (!db) return res.status(500).json({ success: false, message: 'DB no disponible' });

    let oid;
    try { oid = new ObjectId(req.params.id); } catch {
      return res.status(400).json({ success: false, message: 'ID inválido' });
    }

    const { resolucion, notaProcesamiento, fechaVenta, fechaInstalacion } = req.body;
    const resoluciones = ['Venta Completada', 'Venta Pendiente'];
    if (!resoluciones.includes(resolucion)) {
      return res.status(400).json({ success: false, message: 'Resolución inválida' });
    }

    const update = {
      resolucion,
      resueltoEn:   new Date(),
      actualizadoEn: new Date(),
    };
    if (notaProcesamiento !== undefined) update.notaProcesamiento = notaProcesamiento;
    if (fechaVenta)       update.fechaVenta       = fechaVenta;
    if (fechaInstalacion) update.fechaInstalacion = fechaInstalacion;

    const result = await db.collection('pre_leads').updateOne(
      { _id: oid },
      { $set: update }
    );

    if (result.matchedCount === 0) {
      return res.status(404).json({ success: false, message: 'Lead no encontrado' });
    }

    res.json({ success: true, resolucion });
  } catch (err) {
    console.error('[PRE-LEADS] PUT resolver error:', err);
    res.status(500).json({ success: false, message: 'Error interno del servidor' });
  }
});

// POST /api/pre-leads/:id/images — subir imagen a Cloudinary
router.post('/:id/images', protect, imgUpload.single('image'), async (req, res) => {
  try {
    if (!isProcesamiento(req.user.role)) {
      return res.status(403).json({ success: false, message: 'Acceso denegado' });
    }

    const db = getDb();
    if (!db) return res.status(500).json({ success: false, message: 'DB no disponible' });

    let oid;
    try { oid = new ObjectId(req.params.id); } catch {
      return res.status(400).json({ success: false, message: 'ID inválido' });
    }

    if (!req.file) {
      return res.status(400).json({ success: false, message: 'No se recibió imagen' });
    }

    const cloudinary = require('cloudinary').v2;
    const url = await new Promise((resolve, reject) => {
      const stream = cloudinary.uploader.upload_stream(
        { folder: 'dashboard/lead-images', resource_type: 'image' },
        (err, result) => err ? reject(err) : resolve(result.secure_url)
      );
      stream.end(req.file.buffer);
    });

    await db.collection('pre_leads').updateOne(
      { _id: oid },
      { $push: { imagenes: url }, $set: { actualizadoEn: new Date() } }
    );

    res.json({ success: true, url });
  } catch (err) {
    console.error('[PRE-LEADS] Image upload error:', err);
    res.status(500).json({ success: false, message: 'Error al subir imagen' });
  }
});

module.exports = router;

File: CRM_PYTHON/routers/pre_leads.py
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File
from pydantic import BaseModel
from database_mysql import AsyncSessionLocal
from sqlalchemy import text
from deps import current_user
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional
import random, re, aiofiles, datetime as _dt, json

def _utcnow() -> datetime:
    """UTC naive (reemplazo de _utcnow() deprecado en Python 3.12+)."""
    return datetime.now(timezone.utc).replace(tzinfo=None)

router = APIRouter(prefix="/api/pre-leads", tags=["Pre-Leads"])


def _is_procesamiento(role: str) -> bool:
    r = str(role or "").lower().strip()
    return r in ("admin", "administrador", "administrator") or r.startswith("procesamiento")


async def _generar_lead_id(s) -> str:
    for _ in range(20):
        id_ = str(random.randint(10000, 99999))
        r = await s.execute(
            text("SELECT id FROM pre_leads WHERE lead_id = :lid LIMIT 1"), {"lid": id_}
        )
        if not r.first():
            return id_
    return str(random.randint(10000, 99999))


def _fmt(row) -> dict:
    d = dict(row)
    d["_id"] = str(d.get("id", ""))
    # Parse JSON images
    imgs = d.get("images")
    if isinstance(imgs, str):
        try: d["images"] = json.loads(imgs)
        except (ValueError, TypeError): d["images"] = []
    elif imgs is None:
        d["images"] = []
    # Normalize dates to strings
    for col in ("fecha_nacimiento", "fecha_venta", "fecha_instalacion", "resuelto_en", "created_at", "updated_at"):
        v = d.get(col)
        if v is not None:
            d[col] = str(v)
    return d


def _parse_date(s: str) -> Optional[str]:
    if not s:
        return None
    s = s.strip()
    # Already YYYY-MM-DD
    if re.match(r"^\d{4}-\d{2}-\d{2}$", s):
        return s
    # DD/MM/YYYY
    m = re.match(r"^(\d{1,2})/(\d{1,2})/(\d{4})$", s)
    if m:
        return f"{m.group(3)}-{int(m.group(2)):02d}-{int(m.group(1)):02d}"
    return None


class PreLeadBody(BaseModel):
    nombre: str
    correo: str
    phone1: str
    phone2: Optional[str] = ""
    direccion: str
    fechaNacimiento: str
    servicio: str
    mercado: Optional[str] = ""
    nota: Optional[str] = ""
    agenteUsername: Optional[str] = ""
    agenteName: Optional[str] = ""


class UpdateLeadBody(BaseModel):
    status: Optional[str] = None
    notaProcesamiento: Optional[str] = None
    fechaVenta: Optional[str] = None
    fechaInstalacion: Optional[str] = None
    nombre: Optional[str] = None
    correo: Optional[str] = None
    phone1: Optional[str] = None
    phone2: Optional[str] = None
    direccion: Optional[str] = None
    fechaNacimiento: Optional[str] = None
    servicio: Optional[str] = None
    mercado: Optional[str] = None
    nota: Optional[str] = None


class ResolverBody(BaseModel):
    resolucion: str
    notaProcesamiento: Optional[str] = None
    fechaVenta: Optional[str] = None
    fechaInstalacion: Optional[str] = None


@router.post("/")
async def create_pre_lead(body: PreLeadBody, user: dict = Depends(current_user)):
    if not all([body.nombre, body.phone1, body.direccion, body.servicio, body.fechaNacimiento, body.correo]):
        raise HTTPException(400, "Faltan campos obligatorios")
    now = _utcnow()
    fn_date = _parse_date(body.fechaNacimiento)

    async with AsyncSessionLocal() as s:
        lead_id = await _generar_lead_id(s)
        r = await s.execute(text("""
            INSERT INTO pre_leads
                (lead_id, nombre, correo, phone1, phone2, direccion, fecha_nacimiento,
                 servicio, mercado, nota, agente_username, agente_name,
                 status, nota_procesamiento, created_at, updated_at)
            VALUES
                (:lid, :nombre, :correo, :phone1, :phone2, :dir, :fn,
                 :servicio, :mercado, :nota, :au, :an,
                 NULL, '', :now, :now)
        """), {
            "lid": lead_id,
            "nombre":   body.nombre.strip(),
            "correo":   body.correo.strip(),
            "phone1":   body.phone1.strip(),
            "phone2":   (body.phone2 or "").strip(),
            "dir":      body.direccion.strip(),
            "fn":       fn_date,
            "servicio": body.servicio.strip(),
            "mercado":  (body.mercado or "").strip(),
            "nota":     (body.nota or "").strip(),
            "au":       user.get("username") or body.agenteUsername,
            "an":       user.get("name") or body.agenteName,
            "now":      now,
        })
        await s.commit()
        new_id = r.lastrowid
        row = await s.execute(text("SELECT * FROM pre_leads WHERE id = :id"), {"id": new_id})
        doc = _fmt(row.mappings().first())

    return {"success": True, "lead": doc}


@router.get("/mis-leads")
async def mis_leads(user: dict = Depends(current_user)):
    async with AsyncSessionLocal() as s:
        r = await s.execute(text("""
            SELECT * FROM pre_leads WHERE agente_username = :u ORDER BY created_at DESC
        """), {"u": user["username"]})
        leads = [_fmt(row) for row in r.mappings().all()]
    return {"success": True, "leads": leads}


@router.get("")
async def list_pre_leads(user: dict = Depends(current_user)):
    if not _is_procesamiento(user.get("role", "")):
        raise HTTPException(403, "Acceso denegado")
    role = user.get("role", "").lower()

    where = "1=1"
    params: dict = {}
    if "icon" in role:
        where = "LOWER(mercado) = 'icon'"
    elif "bamo" in role:
        where = "LOWER(mercado) = 'bamo'"

    async with AsyncSessionLocal() as s:
        r = await s.execute(
            text(f"SELECT * FROM pre_leads WHERE {where} ORDER BY created_at DESC"), params
        )
        leads = [_fmt(row) for row in r.mappings().all()]
    return {"success": True, "leads": leads}


@router.put("/{lead_id}/resolver")
async def resolver(lead_id: str, body: ResolverBody, user: dict = Depends(current_user)):
    if not _is_procesamiento(user.get("role", "")):
        raise HTTPException(403, "Acceso denegado")
    if body.resolucion not in ("Venta Completada", "Venta Pendiente"):
        raise HTTPException(400, "Resolución inválida")
    try:
        mid = int(lead_id)
    except ValueError:
        raise HTTPException(400, "ID inválido")

    now = _utcnow()
    sets = ["resolucion = :res", "resuelto_en = :now", "updated_at = :now"]
    params: dict = {"res": body.resolucion, "now": now, "id": mid}
    if body.notaProcesamiento is not None:
        sets.append("nota_procesamiento = :np")
        params["np"] = body.notaProcesamiento
    if body.fechaVenta:
        sets.append("fecha_venta = :fv")
        params["fv"] = _parse_date(body.fechaVenta)
    if body.fechaInstalacion:
        sets.append("fecha_instalacion = :fi")
        params["fi"] = _parse_date(body.fechaInstalacion)

    async with AsyncSessionLocal() as s:
        r = await s.execute(
            text(f"UPDATE pre_leads SET {', '.join(sets)} WHERE id = :id"), params
        )
        await s.commit()
        if r.rowcount == 0:
            raise HTTPException(404, "Lead no encontrado")

    return {"success": True, "resolucion": body.resolucion}


_IMAGES_DIR = Path(__file__).parent.parent.parent / "uploads" / "pre_leads"
_IMAGES_DIR.mkdir(parents=True, exist_ok=True)
_MAX_IMG_BYTES = 20 * 1024 * 1024


@router.post("/{lead_id}/images")
async def upload_pre_lead_image(
    lead_id: str, image: UploadFile = File(...), user: dict = Depends(current_user)
):
    try:
        mid = int(lead_id)
    except ValueError:
        raise HTTPException(400, "ID inválido")

    data = await image.read()
    if len(data) > _MAX_IMG_BYTES:
        raise HTTPException(413, "Imagen demasiado grande (máx 20 MB)")

    ext      = Path(image.filename or "img.jpg").suffix.lower() or ".jpg"
    ts       = int(_utcnow().timestamp() * 1000)
    filename = f"{lead_id}_{ts}{ext}"
    dest     = _IMAGES_DIR / filename

    async with aiofiles.open(dest, "wb") as f:
        await f.write(data)

    url = f"/uploads/pre_leads/{filename}"
    new_img = {"url": url, "filename": filename, "uploadedAt": str(_utcnow())}

    async with AsyncSessionLocal() as s:
        # Fetch existing images JSON and append
        row = await s.execute(
            text("SELECT images FROM pre_leads WHERE id = :id LIMIT 1"), {"id": mid}
        )
        r = row.mappings().first()
        if not r:
            raise HTTPException(404, "Lead no encontrado")
        imgs = r.get("images")
        if isinstance(imgs, str):
            try: imgs = json.loads(imgs)
            except (ValueError, TypeError): imgs = []
        imgs = imgs or []
        imgs.append(new_img)
        await s.execute(text("""
            UPDATE pre_leads SET images = :imgs, updated_at = :now WHERE id = :id
        """), {"imgs": json.dumps(imgs), "now": _utcnow(), "id": mid})
        await s.commit()

    return {"success": True, "url": url, "filename": filename}


@router.put("/{lead_id}")
async def update_pre_lead(lead_id: str, body: UpdateLeadBody, user: dict = Depends(current_user)):
    if not _is_procesamiento(user.get("role", "")):
        raise HTTPException(403, "Acceso denegado")
    try:
        mid = int(lead_id)
    except ValueError:
        raise HTTPException(400, "ID inválido")

    col_map = {
        "status":            "status",
        "notaProcesamiento": "nota_procesamiento",
        "fechaVenta":        "fecha_venta",
        "fechaInstalacion":  "fecha_instalacion",
        "nombre":            "nombre",
        "correo":            "correo",
        "phone1":            "phone1",
        "phone2":            "phone2",
        "direccion":         "direccion",
        "fechaNacimiento":   "fecha_nacimiento",
        "servicio":          "servicio",
        "mercado":           "mercado",
        "nota":              "nota",
    }
    date_cols = {"fechaVenta", "fechaInstalacion", "fechaNacimiento"}

    sets = []
    params: dict = {"id": mid, "now": _utcnow()}
    for field, col in col_map.items():
        val = getattr(body, field, None)
        if val is not None:
            if field in date_cols:
                val = _parse_date(str(val))
            else:
                val = str(val).strip()
            sets.append(f"{col} = :{field}")
            params[field] = val

    if not sets:
        raise HTTPException(400, "Sin campos para actualizar")
    sets.append("updated_at = :now")

    async with AsyncSessionLocal() as s:
        r = await s.execute(
            text(f"UPDATE pre_leads SET {', '.join(sets)} WHERE id = :id"), params
        )
        await s.commit()
        if r.rowcount == 0:
            raise HTTPException(404, "Lead no encontrado")

    return {"success": True}
